File: Algorithm/src/main/java/com/zhangdd/search/FibonacciSearch.java
package com.zhangdd.search;

import java.util.Arrays;

/**
 * @author zhangdd on 2021/11/7
 */
public class FibonacciSearch {

    /**
     * 斐波那契队列长度
     */
    private static int maxLength = 20;

    public static void main(String[] args) {

        int arr[] = {1, 2, 3, 5, 8, 8, 10, 18};

        int searchIndex = fibonacciSearch(arr, 3);
        System.out.println(searchIndex);
    }


    /**
     * 构造一个斐波那契队列
     */
    private static int[] buildFibonacciArr() {
        int first = 1;
        int second = 1;
        int[] result = new int[maxLength];
        result[0] = first;
        result[1] = second;

        for (int i = 2; i < maxLength; i++) {
            result[i] = result[i - 1] + result[i - 2];
        }

        return result;
    }

    private static int fibonacciSearch(int[] arr, int findVal) {

        int left = 0;
        int right = arr.length - 1;
        int mid = 0;

        int arrSize = arr.length;
        int[] fibonacciArr = buildFibonacciArr();
        int fibonacciArrIndex = 0;
        while (arrSize > fibonacciArr[fibonacciArrIndex]) {
            fibonacciArrIndex++;
        }
        int[] temp = Arrays.copyOf(arr, fibonacciArr[fibonacciArrIndex]);
        for (int i = arr.length; i < temp.length; i++) {
            temp[i] = arr[arrSize - 1];
        }

        while (left <= right) {
            mid = left + fibonacciArr[fibonacciArrIndex - 1] - 1;

            if (findVal < temp[mid]) {
                right = mid - 1;
                fibonacciArrIndex--;
            } else if (findVal > temp[mid]) {
                left = mid + 1;
                fibonacciArrIndex -= 2;
            } else {

                //这里找到了，返回较小的一个下标
                return Math.min(mid, right);
            }

        }

        return -1;
    }
}

File: python/Algorithm/search/FibonacciSearch.py
#!/usr/bin/env python3
"""
@author: zhangdd on 2022/2/4

斐波那契查找
"""
MAX_LENGTH = 20


def build_fibonacci_arr():
    fibonacci_arr = [1, 1]
    for i in range(2, MAX_LENGTH):
        fibonacci_arr.append(fibonacci_arr[i - 1] + fibonacci_arr[i - 2])
    return fibonacci_arr


def fibonacci_search(arr, value):
    left = 0
    right = len(arr) - 1
    fibonacci_index = 0

    arr_length = len(arr)
    fibonacci_arr = build_fibonacci_arr()

    gap = 0
    for item in fibonacci_arr:
        if arr_length > item:
            fibonacci_index += 1
            continue
        else:
            gap = item - arr_length
            break
    for i in range(gap):
        arr.append(arr[len(arr) - 1])

    while left <= right:
        middle = left + fibonacci_arr[fibonacci_index - 1] - 1
        if value < arr[middle]:
            right = middle - 1
            fibonacci_index -= 1
        elif value > arr[middle]:
            left = middle + 1
            fibonacci_index -= 2
        else:
            return min(middle, right)

    return None


if __name__ == '__main__':
    arr = [1, 2, 3, 5, 8, 8, 10, 18, 19]
    index = fibonacci_search(arr, 20)
    print(index)
